deployments.go

```diff
@@ -58,25 +59,43 @@ func (s *deployerService) deploy(ctx context.Context, db *gorm.DB, opts DeployOp
 		Slug:        slug,
 		Description: desc,
 	})
-
 	return r, err
 }
 
+func (s *deployerService) deployInTransaction(ctx context.Context, stream scheduler.StatusStream, opts DeployOpts) (*Release, error) {
+	tx := s.db.Begin()
+	r, err := s.createRelease(ctx, tx, stream, opts)
+	if err != nil {
+		tx.Rollback()
+		return r, err
+	}
+	if err := tx.Commit().Error; err != nil {
+		return r, err
+	}
+	return r, s.releases.Release(ctx, r, stream)
+}
+
 // Deploy is a thin wrapper around deploy to that adds the error to the
 // jsonmessage stream.
-func (s *deployerService) Deploy(ctx context.Context, db *gorm.DB, opts DeployOpts) (*Release, error) {
+func (s *deployerService) Deploy(ctx context.Context, opts DeployOpts) (*Release, error) {
 	var msg jsonmessage.JSONMessage
 
-	r, err := s.deploy(ctx, db, opts)
+	var stream scheduler.StatusStream
+	if opts.Stream {
+		stream = scheduler.NewJSONMessageStream(opts.Output)
+	}
+
+	r, err := s.deployInTransaction(ctx, stream, opts)
 	if err != nil {
 		msg = newJSONMessageError(err)
-	} else {
+	} else if stream == nil {
 		msg = jsonmessage.JSONMessage{Status: fmt.Sprintf("Status: Created new release v%d for %s", r.Version, r.App.Name)}
```

Comment:
ok, new messaging looks like:

no stream:


```
$ ./build/emp deploy remind101/acme-inc:master
master: Pulling from remind101/acme-inc

420890c9e918: Already exists
a3ed95caeb02: Already exists
1e14aab0083d: Already exists
4393485f2bb3: Already exists
a2a051cf14f7: Already exists
1f9177248efa: Already exists
Digest: sha256:d05682347b56191a950f55ab06f11a6c6e34041921af32281b01f4a6a9a0f75f
Status: Image is up to date for remind101/acme-inc:master
Status: Created new release v10 for acme-inc
Status: Finished processing events for release v10 of acme-inc
```

streaming:

```
$ ./build/emp deploy remind101/acme-inc:master -s
master: Pulling from remind101/acme-inc

420890c9e918: Already exists
a3ed95caeb02: Already exists
1e14aab0083d: Already exists
4393485f2bb3: Already exists
a2a051cf14f7: Already exists
1f9177248efa: Already exists
Digest: sha256:d05682347b56191a950f55ab06f11a6c6e34041921af32281b01f4a6a9a0f75f
Status: Image is up to date for remind101/acme-inc:master
Status: Created new release v11 for acme-inc
Status: Created cloudformation template: https://empire-0b96756d-templatebucket-fzr5jqii49z4.s3.amazonaws.com/acme-inc/04a75f65-c7c4-4e36-943b-0a09614db1b4/5faf9e742e482f0e66c768d1f134986a93795615
Status: Waiting for existing stack operation to complete
Status: Stack update submitted
Status: Waiting for stack update to complete
Status: Stack update complete
Status: Finished processing events for release v11 of acme-inc
```

I added the "Finished..." message to each since after we create the release, we submit to the scheduler, so without it, it might feel like there is an unknown lag.